File: docs/math/code/inverse/inverse-2.py
# --8<-- [start:core]
# Binary exponentiation.
def pow_mod(a, b, m):
    res = 1
    po = a
    while b:
        if b & 1:
            res = (res * po) % m
        po = (po * po) % m
        b >>= 1
    return res


# Returns the modular inverse of a prime modulo p.
def inverse(a, p):
    return pow_mod(a, p - 2, p)


# --8<-- [end:core]

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        a, p = map(int, input().split())
        print(inverse(a, p))

File: docs/math/code/inverse/inverse-2.cpp
#include <iostream>

// --8<-- [start:core]
// Binary exponentiation.
long long pow(long long a, long long b, long long m) {
  long long res = 1, po = a;
  for (; b; b >>= 1) {
    if (b & 1) res = res * po % m;
    po = po * po % m;
  }
  return res;
}

// Returns the modular inverse of a prime modulo p.
long long inverse(long long a, long long p) { return pow(a, p - 2, p); }

// --8<-- [end:core]
int main() {
  long long t;
  std::cin >> t;
  for (; t; --t) {
    long long a, p;
    std::cin >> a >> p;
    std::cout << inverse(a, p) << std::endl;
  }
}
